File: nikkake_kmp/shared/src/commonMain/kotlin/com/myapplication/common/ui/components/Ui.kt
package com.myapplication.common.ui.components

import androidx.compose.foundation.BorderStroke
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.AlertDialog
import androidx.compose.material.Button
import androidx.compose.material.ButtonDefaults
import androidx.compose.material.CircularProgressIndicator
import androidx.compose.material.Text
import androidx.compose.material.TextButton
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.myapplication.common.ui.theme.LocalPalette
import com.myapplication.common.ui.theme.Radii
import com.myapplication.common.ui.theme.Spacing

/**
 * 画面をまたいで使う小さな部品。
 * testTag ではなく contentDescription を使っているのは、
 * Compose UI Test と実際のアクセシビリティの両方で同じ識別子が効くため。
 */

fun Modifier.tag(id: String): Modifier = semantics { contentDescription = id }

@Composable
fun AppCard(
    modifier: Modifier = Modifier,
    color: Color? = null,
    padding: PaddingValues = PaddingValues(Spacing.md),
    content: @Composable () -> Unit,
) {
    val palette = LocalPalette.current

    Box(
        modifier = modifier
            .fillMaxWidth()
            .clip(RoundedCornerShape(Radii.lg))
            .background(color ?: palette.surface)
            .border(1.dp, palette.border, RoundedCornerShape(Radii.lg))
            .padding(padding),
    ) {
        content()
    }
}

@Composable
fun SectionTitle(text: String, modifier: Modifier = Modifier) {
    Text(
        text = text,
        modifier = modifier.padding(bottom = Spacing.md),
        fontSize = 18.sp,
        fontWeight = FontWeight.Bold,
        color = LocalPalette.current.textPrimary,
    )
}

@Composable
fun FieldLabel(text: String) {
    Text(
        text = text,
        modifier = Modifier.padding(bottom = Spacing.sm),
        fontSize = 13.sp,
        fontWeight = FontWeight.SemiBold,
        color = LocalPalette.current.textSecondary,
    )
}

enum class ButtonVariant { PRIMARY, SECONDARY, GHOST, DANGER }

@Composable
fun AppButton(
    label: String,
    onClick: () -> Unit,
    modifier: Modifier = Modifier,
    variant: ButtonVariant = ButtonVariant.PRIMARY,
    enabled: Boolean = true,
    loading: Boolean = false,
) {
    val palette = LocalPalette.current

    val background = when (variant) {
        ButtonVariant.PRIMARY -> palette.primary
        ButtonVariant.SECONDARY -> palette.surfaceElevated
        ButtonVariant.GHOST, ButtonVariant.DANGER -> Color.Transparent
    }
    val foreground = when (variant) {
        ButtonVariant.PRIMARY -> Color.White
        ButtonVariant.SECONDARY -> palette.textPrimary
        ButtonVariant.GHOST -> palette.primaryLight
        ButtonVariant.DANGER -> palette.error
    }
    val border = when (variant) {
        ButtonVariant.SECONDARY -> BorderStroke(1.dp, palette.border)
        ButtonVariant.DANGER -> BorderStroke(1.dp, palette.error)
        else -> null
    }

    Button(
        onClick = onClick,
        modifier = modifier.fillMaxWidth().height(48.dp),
        enabled = enabled && !loading,
        elevation = null,
        shape = RoundedCornerShape(Radii.sm),
        border = border,
        colors = ButtonDefaults.buttonColors(
            backgroundColor = background,
            contentColor = foreground,
            disabledBackgroundColor = background.copy(alpha = 0.4f),
            disabledContentColor = foreground.copy(alpha = 0.6f),
        ),
    ) {
        if (loading) {
            CircularProgressIndicator(color = foreground, strokeWidth = 2.dp, modifier = Modifier.height(20.dp))
        } else {
            Text(label, fontWeight = FontWeight.Bold, fontSize = 15.sp)
        }
    }
}

@Composable
fun EmptyState(
    icon: String,
    title: String,
    message: String,
    modifier: Modifier = Modifier,
    action: (@Composable () -> Unit)? = null,
) {
    val palette = LocalPalette.current

    Column(
        modifier = modifier.fillMaxWidth().padding(Spacing.xl),
        horizontalAlignment = Alignment.CenterHorizontally,
    ) {
        Text(icon, fontSize = 56.sp)
        Text(
            title,
            modifier = Modifier.padding(top = Spacing.md, bottom = Spacing.sm),
            fontSize = 18.sp,
            fontWeight = FontWeight.Bold,
            color = palette.textPrimary,
        )
        Text(message, textAlign = TextAlign.Center, color = palette.textSecondary, fontSize = 14.sp)
        if (action != null) {
            Box(Modifier.padding(top = Spacing.lg)) { action() }
        }
    }
}

@Composable
fun ErrorText(message: String?) {
    if (message.isNullOrEmpty()) return

    Text(
        text = message,
        modifier = Modifier.tag("error-message").padding(bottom = Spacing.md),
        color = LocalPalette.current.error,
        fontSize = 14.sp,
    )
}

@Composable
fun SelectableChip(
    label: String,
    selected: Boolean,
    onClick: () -> Unit,
    modifier: Modifier = Modifier,
) {
    val palette = LocalPalette.current

    Box(
        modifier = modifier
            .clip(RoundedCornerShape(Radii.pill))
            .background(if (selected) palette.primary else palette.surface)
            .border(
                1.dp,
                if (selected) palette.primary else palette.border,
                RoundedCornerShape(Radii.pill),
            )
            .clickable(onClick = onClick)
            .padding(horizontal = Spacing.md, vertical = Spacing.sm),
    ) {
        Text(
            label,
            color = if (selected) Color.White else palette.textPrimary,
            fontSize = 13.sp,
            fontWeight = if (selected) FontWeight.Bold else FontWeight.Normal,
        )
    }
}

@Composable
fun StatBlock(label: String, value: String, valueTag: String, modifier: Modifier = Modifier) {
    val palette = LocalPalette.current

    Column(modifier = modifier, horizontalAlignment = Alignment.CenterHorizontally) {
        Text(label, fontSize = 11.sp, color = palette.textSecondary)
        Text(
            value,
            modifier = Modifier.tag(valueTag).padding(top = Spacing.xs),
            fontSize = 16.sp,
            fontWeight = FontWeight.Bold,
            color = palette.textPrimary,
        )
    }
}

@Composable
fun LabeledRow(label: String, value: String, valueTag: String) {
    val palette = LocalPalette.current

    Row(
        modifier = Modifier.fillMaxWidth().padding(vertical = Spacing.sm),
        horizontalArrangement = Arrangement.SpaceBetween,
        verticalAlignment = Alignment.CenterVertically,
    ) {
        Text(label, fontSize = 14.sp, color = palette.textSecondary)
        Text(
            value,
            modifier = Modifier.tag(valueTag),
            fontSize = 15.sp,
            fontWeight = FontWeight.Bold,
            color = palette.textPrimary,
        )
    }
}

/**
 * 削除など取り返しのつかない操作の確認。
 * プラットフォームのネイティブダイアログではなくComposeで描くのは、
 * 3実装で同じ文言・同じ導線にでき、UIテストからも同じ手順で操作できるため。
 */
@Composable
fun ConfirmDialog(
    title: String,
    message: String,
    confirmLabel: String,
    onConfirm: () -> Unit,
    onDismiss: () -> Unit,
) {
    val palette = LocalPalette.current

    AlertDialog(
        onDismissRequest = onDismiss,
        backgroundColor = palette.surfaceElevated,
        title = { Text(title, color = palette.textPrimary) },
        text = { Text(message, color = palette.textSecondary) },
        confirmButton = {
            TextButton(onClick = onConfirm, modifier = Modifier.tag("confirm-action")) {
                Text(confirmLabel, color = palette.error)
            }
        },
        dismissButton = {
            TextButton(onClick = onDismiss) {
                Text("キャンセル", color = palette.textSecondary)
            }
        },
    )
}

File: nikkake_flutter/lib/widgets/ui.dart
import 'package:flutter/material.dart';
import '../constants/colors.dart';

/// 画面をまたいで使う小さな部品。
/// 3実装(RN/Flutter/KMP)で見た目を揃える必要があるので、
/// 余白・角丸・色はここに集約して各画面で個別に決めない。

class Spacing {
  static const xs = 4.0;
  static const sm = 8.0;
  static const md = 16.0;
  static const lg = 24.0;
  static const xl = 32.0;
}

class Radii {
  static const sm = 8.0;
  static const md = 12.0;
  static const lg = 16.0;
  static const pill = 999.0;
}

/// '#RRGGBB' 形式の文字列をColorへ。DBに色をhexで保存しているため
Color hexToColor(String hex, {Color fallback = AppColors.darkPrimary}) {
  final cleaned = hex.replaceAll('#', '');
  if (cleaned.length != 6) return fallback;
  final value = int.tryParse(cleaned, radix: 16);
  return value == null ? fallback : Color(0xFF000000 | value);
}

class AppCard extends StatelessWidget {
  final Widget child;
  final EdgeInsetsGeometry? padding;
  final Color? color;

  const AppCard({super.key, required this.child, this.padding, this.color});

  @override
  Widget build(BuildContext context) {
    return Container(
      width: double.infinity,
      padding: padding ?? const EdgeInsets.all(Spacing.md),
      decoration: BoxDecoration(
        color: color ?? AppColors.darkSurface,
        borderRadius: BorderRadius.circular(Radii.lg),
        border: Border.all(color: AppColors.darkBorder),
      ),
      child: child,
    );
  }
}

class SectionTitle extends StatelessWidget {
  final String text;
  final EdgeInsetsGeometry? padding;

  const SectionTitle(this.text, {super.key, this.padding});

  @override
  Widget build(BuildContext context) {
    return Padding(
      padding: padding ?? const EdgeInsets.only(bottom: Spacing.md),
      child: Text(
        text,
        style: const TextStyle(
          fontSize: 18,
          fontWeight: FontWeight.bold,
          color: AppColors.darkTextPrimary,
        ),
      ),
    );
  }
}

enum AppButtonVariant { primary, secondary, ghost, danger }

class AppButton extends StatelessWidget {
  final String label;
  final VoidCallback? onPressed;
  final AppButtonVariant variant;
  final bool loading;
  final bool expand;

  const AppButton({
    super.key,
    required this.label,
    this.onPressed,
    this.variant = AppButtonVariant.primary,
    this.loading = false,
    this.expand = true,
  });

  @override
  Widget build(BuildContext context) {
    final disabled = loading || onPressed == null;

    final (bg, fg, border) = switch (variant) {
      AppButtonVariant.primary => (AppColors.darkPrimary, Colors.white, null),
      AppButtonVariant.secondary => (
          AppColors.darkSurfaceElevated,
          AppColors.darkTextPrimary,
          AppColors.darkBorder
        ),
      AppButtonVariant.ghost => (Colors.transparent, AppColors.darkPrimaryLight, null),
      AppButtonVariant.danger => (Colors.transparent, AppColors.darkError, AppColors.darkError),
    };

    return Opacity(
      opacity: disabled ? 0.4 : 1,
      child: SizedBox(
        width: expand ? double.infinity : null,
        height: 48,
        child: ElevatedButton(
          onPressed: disabled ? null : onPressed,
          style: ElevatedButton.styleFrom(
            backgroundColor: bg,
            foregroundColor: fg,
            disabledBackgroundColor: bg,
            disabledForegroundColor: fg,
            elevation: 0,
            shape: RoundedRectangleBorder(
              borderRadius: BorderRadius.circular(Radii.sm),
              side: border == null ? BorderSide.none : BorderSide(color: border),
            ),
          ),
          child: loading
              ? const SizedBox(
                  width: 20,
                  height: 20,
                  child: CircularProgressIndicator(strokeWidth: 2, color: Colors.white),
                )
              : Text(label, style: const TextStyle(fontWeight: FontWeight.bold, fontSize: 15)),
        ),
      ),
    );
  }
}

class EmptyState extends StatelessWidget {
  final String icon;
  final String title;
  final String message;
  final Widget? action;

  const EmptyState({
    super.key,
    required this.icon,
    required this.title,
    required this.message,
    this.action,
  });

  @override
  Widget build(BuildContext context) {
    return Padding(
      padding: const EdgeInsets.all(Spacing.xl),
      child: Column(
        mainAxisSize: MainAxisSize.min,
        children: [
          Text(icon, style: const TextStyle(fontSize: 56)),
          const SizedBox(height: Spacing.md),
          Text(
            title,
            style: const TextStyle(
              fontSize: 18,
              fontWeight: FontWeight.bold,
              color: AppColors.darkTextPrimary,
            ),
          ),
          const SizedBox(height: Spacing.sm),
          Text(
            message,
            textAlign: TextAlign.center,
            style: const TextStyle(color: AppColors.darkTextSecondary, height: 1.5),
          ),
          if (action != null) ...[const SizedBox(height: Spacing.lg), action!],
        ],
      ),
    );
  }
}

class ErrorText extends StatelessWidget {
  final String? message;

  const ErrorText(this.message, {super.key});

  @override
  Widget build(BuildContext context) {
    if (message == null || message!.isEmpty) return const SizedBox.shrink();

    return Padding(
      padding: const EdgeInsets.only(bottom: Spacing.md),
      child: Text(
        message!,
        key: const Key('error-message'),
        style: const TextStyle(color: AppColors.darkError, fontSize: 14),
      ),
    );
  }
}

class SelectableChip extends StatelessWidget {
  final String label;
  final bool selected;
  final VoidCallback onTap;

  const SelectableChip({
    super.key,
    required this.label,
    required this.selected,
    required this.onTap,
  });

  @override
  Widget build(BuildContext context) {
    return GestureDetector(
      onTap: onTap,
      child: Container(
        padding: const EdgeInsets.symmetric(horizontal: Spacing.md, vertical: Spacing.sm),
        decoration: BoxDecoration(
          color: selected ? AppColors.darkPrimary : AppColors.darkSurface,
          borderRadius: BorderRadius.circular(Radii.pill),
          border: Border.all(color: selected ? AppColors.darkPrimary : AppColors.darkBorder),
        ),
        child: Text(
          label,
          style: TextStyle(
            color: selected ? Colors.white : AppColors.darkTextPrimary,
            fontSize: 13,
            fontWeight: selected ? FontWeight.bold : FontWeight.normal,
          ),
        ),
      ),
    );
  }
}

/// 削除など取り返しのつかない操作の確認
Future<bool> confirmAction(
  BuildContext context, {
  required String title,
  required String message,
  String confirmLabel = '削除',
}) async {
  final result = await showDialog<bool>(
    context: context,
    builder: (context) => AlertDialog(
      backgroundColor: AppColors.darkSurfaceElevated,
      title: Text(title, style: const TextStyle(color: AppColors.darkTextPrimary)),
      content: Text(message, style: const TextStyle(color: AppColors.darkTextSecondary)),
      actions: [
        TextButton(
          onPressed: () => Navigator.of(context).pop(false),
          child: const Text('キャンセル'),
        ),
        TextButton(
          key: const Key('confirm-action'),
          onPressed: () => Navigator.of(context).pop(true),
          child: Text(confirmLabel, style: const TextStyle(color: AppColors.darkError)),
        ),
      ],
    ),
  );

  return result ?? false;
}
